### PR TITLE
construct layer config to pass parameters from cfg to layers.

### PR DESCRIPTION
**Pull request by yaoyaowd**
This PR introduces intermediate layer config to pass parameters from cfg to layers.

Next step:
1. I will refactor general conv and combine generalconv.py and general_conv.py into one layer.
2. I will refactor some other configurations for graph gym so that configurations are more modularized.

### CODE REVIEW

**Comment by rusty1s** [comment]
Thanks @yaoyaowd. This now makes `GeneralConv` be a lazy operator by design, which might break GraphGym at the moment. Instead, I feel that we should make `edge_dim` an argument of the `GeneralEdgeConvLayer`.

**Comment by yaoyaowd** [comment]
I tested with example bash and they worked fine. Currently there's a sequence of steps to create the GeneralEdgeConv:
`main->model_builder->gnn->layer->generalconv`. They all access the same cfg file and I think it is better to reduce the parameter we need from cfg and then refactor each piece slowly.

**Comment by rusty1s** [comment]
I absolutely agree with you. I'm just of the opinion that lazy ops currently do not work in GraphGym, as the optimizer is initialized before the very first forward pass. This does result in an error, but it certainly leads to lazy `Linear` layers being untrained.

**Comment by yaoyaowd** [comment]
I did a bigger refactor so that there's a intermediate layer_config between gnn and layers.

File: torch_geometric/graphgym/models/layer.py [hunk 1]
```diff
@@ -1,16 +1,67 @@
...
 
-from torch_geometric.graphgym.config import cfg
 from torch_geometric.graphgym.models.act import act_dict
 from torch_geometric.graphgym.contrib.layer.generalconv import (
     GeneralConvLayer, GeneralEdgeConvLayer)
 
 import torch_geometric.graphgym.register as register
 
 
+@dataclass
```
Comment (rusty1s) [inline torch_geometric/graphgym/models/layer.py:16]: Is there a difference between `@dataclass` and simply using `NamedTuple`s?

**Comment by rusty1s** [comment]
This looks better IMO. Thank you. Let's see what @JiaxuanYou thinks.

File: torch_geometric/graphgym/models/layer.py [hunk 2]
```diff
@@ -156,18 +223,21 @@ class MLP(nn.Module):
...
+        dim_inner = layer_config.dim_in \
+            if layer_config.dim_inner is None \
+            else layer_config.dim_inner
+        layer_config.has_bias = True
         layers = []
-        if num_layers > 1:
+        if layer_config.num_layers > 1:
             layers.append(
-                GeneralMultiLayer('linear', num_layers - 1, dim_in, dim_inner,
+                GeneralMultiLayer('linear', layer_config.num_layers - 1,
```
Comment (JiaxuanYou) [inline torch_geometric/graphgym/models/layer.py:235]: E.g., user may not know there is a `layer_config.num_layers` in `layer_config`

File: torch_geometric/graphgym/models/layer.py [hunk 3]
```diff
@@ -156,18 +223,21 @@ class MLP(nn.Module):
         num_layers (int): Number of layers in the stack
         **kwargs (optional): Additional args
     """
-    def __init__(self, dim_in, dim_out, bias=True, dim_inner=None,
-                 num_layers=2, **kwargs):
+    def __init__(self, layer_config: LayerConfig, **kwargs):
```
Comment (JiaxuanYou) [inline torch_geometric/graphgym/models/layer.py:226]: E.g., has_batchnorm does not apply here, but a user may not know that, and still set `has_batchnorm` and expect it to work.

File: torch_geometric/graphgym/models/layer.py (hunk 1 above)
Comment (yaoyaowd) [reply torch_geometric/graphgym/models/layer.py:16]: Data class is a class to store data with attribute while named tuple is a dictionary under the hood.
https://www.geeksforgeeks.org/difference-between-dataclass-vs-namedtuple-vs-object-in-python/

**Comment by yaoyaowd** [comment]
The main thing I want to address is to modularize parameters for different function areas, i.e dataset accesses parameters from data_config, model accesses parameters from model_config, layer accesses parameters from layer_config.

I think issues you described above are both hard problems for a general LayerConfig and the cfg. Where users need to be familiar with cfg and parameters in cfg may not impact the model if it is not config correctly. But I want to make a step forward so that we know what parameters in the cfg goes into layers and then we separate the problem into two smaller problems that can be solved in parallel.
- continue refactor cfg into individual module config.
- refactor layer config to reflect different config for different layers.

**Comment by JiaxuanYou** [comment]
Thanks for explaining. This makes sense to me, and I agree with the whole picture of improving the structure of cfg.

**Comment by yaoyaowd** [comment]
Thanks, I will assume it is a LGTM and merge the code :)

**Pull request merged**